arwen/contexts/storage.go

```diff
@@ -92,44 +104,47 @@ func (context *storageContext) GetStorageUpdates(address []byte) map[string]*vmc
 }
 
 // GetStorage returns the storage data mapped to the given key.
-func (context *storageContext) GetStorage(key []byte) []byte {
+func (context *storageContext) GetStorage(key []byte) ([]byte, bool) {
 	metering := context.host.Metering()
 
+	value, usedCache := context.GetStorageUnmetered(key)
+
 	extraBytes := len(key) - arwen.AddressLen
-	if extraBytes > 0 {
+	gasFlagSet := context.flagUseDifferentGasCostForReadingCachedStorage.IsSet()
+	if (extraBytes > 0 && !gasFlagSet) || (extraBytes > 0 && gasFlagSet && !usedCache) {
```

Comment:
this can be simplified. if (extraBytes > 0) && (!gasFlagSet || !usedCache)

Reply:
nice!